### PR TITLE
Add script to compute vertical coordinate for initial condition

### PR DESCRIPTION
**Pull request by mark-petersen**
Previously, the vertical coordinate (layer depths) was read from a file when generating initial conditions in COMPASS. This adds a script with input parameters and adds it to the COMPASS init sequence. This will be used to create a 64-layer domain for the EC60to30 and others right now.

### CODE REVIEW

**Comment by mark-petersen** [comment]
@vanroekel thanks for providing the script. I only altered formatting, but not content.

**Comment by mark-petersen** [comment]
I added a little plotting routine to the script. Parameters are in bottom right.
vertical_grid
I just made a guess at 10m and 125m for layer min and max.  @vanroekel and @maltrud do you have a recommendation for the new 64-layer grid?

**Comment by mark-petersen** [comment]
Here are the current settings for the 64-layer grid:
```
number layers: 64
bottom depth requested:   5500.00
bottom depth actual:      5647.01
min thickness reqeusted:     2.00
min thickness actual:        1.98
max thickness reqeusted:   200.00
max thickness actual:      194.39
```
vertical_grid

**Comment by maltrud** [comment]
@mark-petersen: Andrew Roberts (@proteanplanet) has generated a 64 level profile.  maybe using the same code, but you should make sure you guys are on the same page.

**Comment by mark-petersen** [comment]
@proteanplanet, please check the layer distribution above. Can you check against or post your 64-layer grid? Did you use the script from @vanroekel or something else?

**Comment by vanroekel** [comment]
I generated the layers for @proteanplanet -- here are the layer thicknesses from that invocation of the script

 1.98312923   2.22519441   2.49685395   2.80181737   3.1442291
   3.52855849   3.95992628   4.44421212   4.98772461   5.59774398
   6.28240812   7.05070601   7.91279578   8.87999147   9.96496251
  11.18181656  12.54628046  14.07564869  15.78925496  17.70826202
  19.855847    22.25743896  24.94034081  27.9336851   31.26825453
  34.9755951   39.08744573  43.63426116  48.64342975  54.13695011
  60.12858277  66.61998041  73.5971205   81.02584967  88.84911568
  96.98450166 105.32498239 113.74241952 122.09422308 130.23325689
138.01922673 145.32973767 152.06928836 158.1743376  163.61443548
168.38947637 172.52432948 176.06218087 179.05781446 181.57172959
183.66541283 185.39799053 186.82412992 187.99284603 188.94713296
189.72400846 190.35491212 190.86625967 191.28003803 191.61442574
191.88437023 192.10210492 192.27760581 192.41898635

**Comment by vanroekel** [comment]
sorry, accidentally clicked closed.  

I was going to say that I don't have the parameters I used anymore.  but your graph looks close to what I post here.

File: testing_and_setup/compass/ocean/global_ocean/config_files/config_initial_state.xml [hunk 1]
```diff
@@ -52,6 +54,14 @@
 		<step executable="gpmetis">
 			<argument flag="graph.info">8</argument>
 		</step>
+		<step executable="./make_vertical_grid.py">
```
Comment (mark-petersen) [inline testing_and_setup/compass/ocean/global_ocean/config_files/config_initial_state.xml]: @xylar, I've only tested the QU240 right now, but this file changes all of them. Is it best to call the python script as an command-line executable here? That's what I'm most familiar with.

**Comment by mark-petersen** [comment]
>I generated the layers for @proteanplanet -- here are the layer thicknesses from that invocation of the script
>1.98312923 2.22519441 2.49685395 2.80181737 3.1442291
3.52855849 3.95992628 4.44421212 4.98772461 5.59774398
6.28240812 7.05070601 7.91279578 8.87999147 9.96496251
11.18181656 12.54628046 14.07564869 15.78925496 17.70826202
19.855847 22.25743896 24.94034081 27.9336851 31.26825453
34.9755951 39.08744573 43.63426116 48.64342975 54.13695011
60.12858277 66.61998041 73.5971205 81.02584967 88.84911568
96.98450166 105.32498239 113.74241952 122.09422308 130.23325689
138.01922673 145.32973767 152.06928836 158.1743376 163.61443548
168.38947637 172.52432948 176.06218087 179.05781446 181.57172959
183.66541283 185.39799053 186.82412992 187.99284603 188.94713296
189.72400846 190.35491212 190.86625967 191.28003803 191.61442574
191.88437023 192.10210492 192.27760581 192.41898635

@vanroekel thanks for the values. This new one looks extremely close:
[  1.97697012   2.1874616    2.42467604   2.69195132   2.99262391
   3.33114193   3.71195067   4.13949159   4.62154066   5.16363904
   5.77354315   6.45899347   7.22993359   8.09738692   9.07344281
  10.17123718  11.40492584  12.79186606  14.34923108  16.09836972
  18.06247351  20.26532146  22.7343817   25.49830612  28.58872035
  32.03856028  35.88239971  40.15328066  44.88357416  50.10222321
  55.83163801  62.08427362  68.85899273  76.13741704  83.87787262
  92.01346542 100.44848303 109.06282554 117.71363083 126.2458942
 134.5039678  142.34350209 149.64357922 156.31506951 162.30477184
 167.59450648 172.19671331 176.14787945 179.50099529 182.31795284
 184.66449925 186.60513888 188.20052864 189.50552546 190.5686493
 191.43188142 192.13087372 192.69560719 193.15107425 193.51786304
 193.81291227 194.05002642 194.2404337  194.3932398 ]

**Comment by proteanplanet** [comment]
@vanroekel I'll run this through my scripts later today and post the results here and on Confluence.

File: testing_and_setup/compass/ocean/scripts/vertical_grid/define_vertical_grid_64_layers.py [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+import subprocess
+
+# Standard 64 layer vertical grid
+subprocess.check_call(['./make_vertical_grid.py', 
+    '--output_file_name', 'vertical_grid.nc',
+    '--num_vert_levels', '64',
+    '--bottom_depth', '5500',
+    '--layer1_thickness', '2',
+    '--maxLayer_thickness', '200',
+    '--plot_vertical_grid'])
```
Comment (xylar) [inline testing_and_setup/compass/ocean/scripts/vertical_grid/define_vertical_grid_64_layers.py]: These kinds of subprocess calls are not working well with the MPI versions of the COMPASS environment.  It would be better to make this a function and call it that way.  In general, calling python scripts from other python scripts with subprocess is not a good practice.

**Comment by mark-petersen** [comment]
OK, I've tested with QU240, QU240wISC, and EC60to30. I am confident the COMPASS a scripts are all working.

File: testing_and_setup/compass/ocean/global_ocean/config_files/config_initial_state.xml (hunk 1 above)
Comment (xylar) [reply testing_and_setup/compass/ocean/global_ocean/config_files/config_initial_state.xml]: Yes, this is fine.

File: testing_and_setup/compass/ocean/scripts/vertical_grid/define_vertical_grid_64_layers.py (hunk 2 above)
Comment (xylar) [reply testing_and_setup/compass/ocean/scripts/vertical_grid/define_vertical_grid_64_layers.py]: @mark-petersen, do you want me to make these changes or do you want to do it?

Comment (mark-petersen) [reply testing_and_setup/compass/ocean/scripts/vertical_grid/define_vertical_grid_64_layers.py]: So it should look like this instead?
```
include make_vertical_grid

make_vertical_grid(arg1, arg2, ...)
```

File: testing_and_setup/compass/ocean/scripts/vertical_grid/make_vertical_grid.py [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+    else:
+        refBottomDepth[:] = botDepth
+        refMidDepth[:] = midDepth
+        refLayerThickness[:] = layerThickness[:nVertLevels]
+        ncfile.close()
+
+    if plot_vertical_grid:
+        import matplotlib
+        import matplotlib.pyplot as plt
+        matplotlib.use('Agg')
```
Comment (xylar) [inline testing_and_setup/compass/ocean/scripts/vertical_grid/make_vertical_grid.py]: Imports like this are considered poor practice.  This should be moved to the top of the script even if it is not always used.  Also, the `use('Agg')` call needs to happen before `plt` gets imported or it won't work right.

File: testing_and_setup/compass/ocean/scripts/vertical_grid/make_vertical_grid.py [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+        plt.grid()
+
+        txt = \
+            'number layers: ' + str(nz) + '\n' + \
+            'bottom depth requested:  ' + '{:8.2f}'.format(bottom_depth) + '\n' +  \
+            'bottom depth actual:     ' + '{:8.2f}'.format(np.amax(botDepth)) + '\n' +  \
+            'min thickness reqeusted: ' + '{:8.2f}'.format(dz1_in) + '\n' + \
+            'min thickness actual:    ' + '{:8.2f}'.format(np.amin(layerThickness[:])) + '\n' + \
+            'max thickness reqeusted: ' + '{:8.2f}'.format(dz2_in) + '\n' + \
+            'max thickness actual:    ' + '{:8.2f}'.format(np.amax(layerThickness[:]))
```
Comment (xylar) [inline testing_and_setup/compass/ocean/scripts/vertical_grid/make_vertical_grid.py]: This is a strange combination for formatting strings.  Generally, `str` addition is not used, and it isn't needed at all when you're using `format` like here.

File: testing_and_setup/compass/ocean/scripts/vertical_grid/define_vertical_grid_64_layers.py (hunk 2 above)
Comment (mark-petersen) [reply testing_and_setup/compass/ocean/scripts/vertical_grid/define_vertical_grid_64_layers.py]: OK, that one is done.

File: testing_and_setup/compass/ocean/scripts/vertical_grid/make_vertical_grid.py [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+
+def create_vertical_grid(
+        bottom_depth,
+        nz,
+        dz1_in,
+        dz2_in,
+        plot_vertical_grid,
+        maxit,
+        epsilon,
+        outFile):
```
Comment (xylar) [inline testing_and_setup/compass/ocean/scripts/vertical_grid/make_vertical_grid.py]: I would define all these arguments with the same default values as in the command-line interface so they are optional.
```
bottom_depth=5000.0,
nz=64,
...
```

File: testing_and_setup/compass/ocean/scripts/vertical_grid/make_vertical_grid.py (hunk 3 above)
Comment (mark-petersen) [reply testing_and_setup/compass/ocean/scripts/vertical_grid/make_vertical_grid.py]: Thanks. That's good to know.

File: testing_and_setup/compass/ocean/scripts/vertical_grid/make_vertical_grid.py (hunk 5 above)
Comment (xylar) [reply testing_and_setup/compass/ocean/scripts/vertical_grid/make_vertical_grid.py]: Also, there should be a docstring documenting this function. It can basically copy the help:
```
"""
This function creates the vertical grid for MPAS-Ocean and writes it to a NetCDF file.

Parameters
----------

bottom_depth : float
    bottom depth for the chosen vertical coordinate [m]

nz : int
    Number of vertical levels for the grid
...
"""
```

File: testing_and_setup/compass/ocean/scripts/vertical_grid/make_vertical_grid.py (hunk 4 above)
Comment (mark-petersen) [reply testing_and_setup/compass/ocean/scripts/vertical_grid/make_vertical_grid.py]: I can replace the str(nz) with something like `'{:d}'.format(nz)`. Is that what you mean? Or was there something else?

File: testing_and_setup/compass/ocean/scripts/vertical_grid/define_vertical_grid_64_layers.py (hunk 2 above)
Comment (xylar) [reply testing_and_setup/compass/ocean/scripts/vertical_grid/define_vertical_grid_64_layers.py]: It should look like this:
```python
#!/usr/bin/env python
from make_vertical_grid import create_vertical_grid


# Standard 64 layer vertical grid
create_vertical_grid(bottom_depth=5500, nz=64, dz1_in=2.0, dz2_in=200.0,
                     plot_vertical_grid=True, maxit=1000, epsilon=1e-2,
                     outfile='vertical_grid.nc')
```

File: testing_and_setup/compass/ocean/scripts/vertical_grid/define_vertical_grid_64_layers.py [hunk 6]
```diff
@@ -1,12 +1,14 @@
...
-    '--bottom_depth', '5500',
-    '--layer1_thickness', '2',
-    '--maxLayer_thickness', '200',
-    '--plot_vertical_grid'])
+make_vertical_grid.create_vertical_grid(
+     5500,            # bottom_depth
+     64,              # num_vert_levels
+     2,               # layer1_thickness
+     200,             # maxLayer_thickness
+     True,            # plot_vertical_grid
```
Comment (xylar) [inline testing_and_setup/compass/ocean/scripts/vertical_grid/define_vertical_grid_64_layers.py]: These comments wouldn't be necessary if you do `bottom_detph=5500`.  If the parameters for `create_vertical_grid` aren't clear, they should be renamed so the comments aren't necessary.

File: testing_and_setup/compass/ocean/scripts/vertical_grid/define_vertical_grid_64_layers.py [hunk 7]
```diff
@@ -1,12 +1,14 @@
...
+     maxit=1000,
+     epsilon=1.0e-2,
+     outFile= 'vertical_grid.nc')# output_file_name
```
Comment (xylar) [inline testing_and_setup/compass/ocean/scripts/vertical_grid/define_vertical_grid_64_layers.py]: In general, comments on lines are discouraged in python.  this one is not needed.

File: testing_and_setup/compass/ocean/scripts/vertical_grid/make_vertical_grid.py (hunk 4 above)
Comment (xylar) [reply testing_and_setup/compass/ocean/scripts/vertical_grid/make_vertical_grid.py]: In this particular case, the pluses might be okay for each line.  In general, there would only be one string with a bunch of curly brackets and one format statement at the end for all the variables:

```python
         txt = 'number layers: {} \n' \
               'bottom depth requested:  {:8.2f}\n'  \
               'bottom depth actual:     {:8.2f}\n'.format(nz, bottom_depth, np.amax(botDepth))
```
In this case, since it's so long, maybe this would work (i.e. a plus sign per line):
```python
        txt = 'number layers: {} \n'.format(nz) + \
              'bottom depth requested:  {:8.2f}\n'.format(bottom_depth) +  \
              'bottom depth actual:     {:8.2f}\n'.format(np.amax(botDepth)) +  \
...
```

Comment (xylar) [reply testing_and_setup/compass/ocean/scripts/vertical_grid/make_vertical_grid.py]: In general, we don't use pluses with strings in python.

**Comment by proteanplanet** [comment]
@vanroekel - An assorted set of comparisons between 60-, 64- and 80-levels:
64-level-grid_1
64-level-grid_2
64-level-grid_3
64-level-grid_4
64-level-grid_5
64-level-grid_6

File: testing_and_setup/compass/ocean/global_ocean/QU240wISC/init/define_vertical_grid.py [hunk 8]
```diff
@@ -0,0 +1 @@
+../../../scripts/vertical_grid/define_vertical_grid_64_layers.py
```
Comment (xylar) [inline testing_and_setup/compass/ocean/global_ocean/QU240wISC/init/define_vertical_grid.py]: I think links like this are needed for a lot more test cases, right?
* CUSP12
* CUSP8
* QU60
* SO60to10wISC
* SOQU60to15 (is this even a configuration we want to keep?)

**Comment by xylar** [comment]
Suggested alternative algorithm:
https://gist.github.com/xylar/9b4506c3e9931ab2b7ecd83f96a00427
In `root_finding_algorithm.py`, I have pulled out only the functions related to creating the vertical grid just so they are easier to follow.

The gist is that we're solving for `delta` in a resolution function:
```
dz = (dz2 - dz1) * np.tanh(-z * np.pi / delta) + dz1
```
where `dz1` is the value of `dz` at `z = 0` and `dz2` is the same at `z --> infinity`, such that `z[nz+1] == -bottom_depth`.  `z` is computed by a cumulative sum, `z[k+1] = z[k] - dz(z[k])`.  As near as I can tell, @vanroekel's algorithm, instead tries to compute `z[k+1] = z[k] - dz(z[k+1])`, but I don't think that is a necessary complication to add to the problem, especially since `dz` should ideally be computed at `z[k+1/2]` for accuracy.

If I follow, @vanroekel's algorithm holds `delta` fixed at `bottom_depth` and iteratively varies `dz2` to attempt to achieve `z[nz+1] == -bottom_depth`.  But the process is hard to follow and doesn't look very robust to me.  The result is also not particularly close to the requested values.  Thus, my suggestion that we move to a new approach.  For reference, here is what my algorithm produces for the 64-layer grid:

vertical_grid_xsad

**Comment by vanroekel** [comment]
@xylar the algorithm is from this paper -- https://www.sciencedirect.com/science/article/abs/pii/S1463500317300434

I hacked together the script to reproduce what they describe.  I'm sure there are much better ways to achieve the same result and I"m perfectly fine switching to something else.  If your algorithm produces similar dz, I'd say we use that instead.

I wrote this script so long ago that I'd have to look at it closely to remember exactly what it does, but a quick look suggests your read is correct, it is a kluged iterative solve.  And I do recall that the algorithm is sensitive to. choices of things like dz1, dz2, and the step size and does fail to converge for some values.  So if your algorithm is more robust and produces similar dz, we should certainly replace what is there now.

**Comment by mark-petersen** [comment]
@vanroekel thanks for the reference. I must admit I found the first version hard to understand. 

@xylar could you push your version, and add a few explanatory comments in the code, similar to your description above? Then I'll finish up any more corrections you mentioned above.

**Comment by xylar** [comment]
@mark-petersen and @vanroekel, I appreciate your flexibility.  I have replaced the script with my version.  I documented each function with a full docstring.  Let me know if that is sufficient or if you would like additional comments.

**Comment by xylar** [comment]
> @xylar the algorithm is from this paper -- https://www.sciencedirect.com/science/article/abs/pii/S1463500317300434

> I hacked together the script to reproduce what they describe.

@vanroekel, I'm kind of glad to know that that wasn't your algorithm.  I realize this is a public forum and the authors may see my comment, but I don't understand why they came up with such a needlessly complicated algorithm.

**Comment by vanroekel** [comment]
I just tested this and it worked fine for me.  Here is a comparison with my original values posted above

Screen Shot 2020-03-31 at 11 23 37 AM

the new script has more resolution in intermediate depths.  I think this is a good improvement.  We'd want more resolution there to better resolve the vertical structure of mesoscale eddies instead of near the bottom.

**Comment by mark-petersen** [comment]
@xylar after today's progress, is it OK if we merge this before #450? Then I can have a big E3SM 'COMPASS only' PR, and take the Fortran PRs in smaller bunches for E3SM. But if you still prefer to wait on this, I'll do it after #450.

**Comment by mark-petersen** [comment]
BTW, I will still address your remaining comments above on this PR.

**Comment by xylar** [comment]
Thought a phone call would be easier...

My concern about merging this is whether it makes it difficult to (temporarily) go back to the 60-level mesh for the EC60to30wISC initial condition that's currently our highest priority.  I guess it shouldn't be hard to do that, though, right?

**Comment by xylar** [comment]
During our meeting today, @maltrud suggested setting the maximum depth to 6000 or 6500 m.  Here is a comparison of 5500 (black) with 6500 (red).

vertical_grid_5500_6500

The lower right panel is just a zoom in on the upper right.

**Comment by mark-petersen** [comment]
Rebased on ocean/develop. Old 60 layer case is still available, but default is now the 64 layer case. I changed the test QU240 to 16 layers for speed.

File: testing_and_setup/compass/ocean/scripts/vertical_grid/define_vertical_grid_64_layers.py [hunk 9]
```diff
@@ -0,0 +1,12 @@
...
+
+from make_vertical_grid import create_vertical_grid
+
+# Standard 64 layer vertical grid
+create_vertical_grid(
+    num_vert_levels=64,
+    bottom_depth=6000,
+    min_layer_thickness=2.,
+    max_layer_thickness=210.,
+    plot_vertical_grid=True,
```
Comment (mark-petersen) [inline testing_and_setup/compass/ocean/scripts/vertical_grid/define_vertical_grid_64_layers.py:11]: Based on the comments at the [meeting today](https://acme-climate.atlassian.net/wiki/spaces/ECG/pages/1307443237/04-01-2020+Cryosphere+Simulation+Campaign?focusedCommentId=1310818552#comment-1310818552) I set the 64-layer bottom depth to 6000m, varying from 2m to 210m.

**Comment by mark-petersen** [comment]
Here is the auto-generated plot:
vertical_grid
The raw numbers are here:
https://gist.github.com/mark-petersen/f4b9617040783d17c1fd65c2f02ee7bb

**Comment by mark-petersen** [comment]
EC60to30 with 64 levels. This is the new plot that is produced at the end of the initial_state step:
initial_state
